src/main/scala/mesosphere/marathon/core/election/CuratorElectionStream.scala

```diff
@@ -0,0 +1,230 @@
+package mesosphere.marathon
+package core.election
+
+import akka.actor.Cancellable
+import akka.stream.scaladsl.{ Source, SourceQueueWithComplete }
+import akka.stream.OverflowStrategy
+import com.typesafe.scalalogging.StrictLogging
+import java.util
+import java.util.Collections
+import java.util.concurrent.{ ExecutorService, TimeUnit }
+import mesosphere.marathon.metrics.{ Metrics, ServiceMetric, Timer }
+import mesosphere.marathon.stream.EnrichedFlow
+import mesosphere.marathon.util.LifeCycledCloseableLike
+import org.apache.curator.framework.CuratorFramework
+import org.apache.curator.framework.api.{ ACLProvider, CuratorWatcher, UnhandledErrorListener }
+import org.apache.curator.framework.imps.CuratorFrameworkState
+import org.apache.curator.framework.recipes.leader.LeaderLatch
+import org.apache.curator.framework.{ AuthInfo, CuratorFrameworkFactory }
+import org.apache.curator.retry.ExponentialBackoffRetry
+import org.apache.zookeeper.{ KeeperException, WatchedEvent, ZooDefs }
+import org.apache.zookeeper.data.ACL
+import scala.concurrent.Future
+import scala.concurrent.duration._
+import scala.concurrent.ExecutionContext
+import scala.util.Try
+
+object CuratorElectionStream extends StrictLogging {
+  /**
+    * Connects to Zookeeper and offers leadership; monitors leader state. Watches for leadership changes (leader
+    * changed, was elected leader, lost leadership), and emits events accordingly.
+    *
+    * Materialized cancellable is used to abdicate leadership; which will do so followed by a closing of the stream.
+    */
+  def apply(
+    clientCloseable: LifeCycledCloseableLike[CuratorFramework],
+    zooKeeperLeaderPath: String,
+    zooKeeperConnectionTimeout: FiniteDuration,
+    hostPort: String,
+    singleThreadExecutor: ExecutorService): Source[LeadershipState, Cancellable] = {
+    val singleThreadEC = ExecutionContext.fromExecutor(singleThreadExecutor)
+    Source.queue[LeadershipState](16, OverflowStrategy.dropHead)
+      .mapMaterializedValue { sq =>
+        val emitterLogic = new CuratorEventEmitter(singleThreadExecutor, clientCloseable, zooKeeperLeaderPath, hostPort, sq)
+        emitterLogic.start()
+        sq.watchCompletion().onComplete { _ => emitterLogic.cancel() }(singleThreadEC)
+        emitterLogic
+      }
+      .initialTimeout(zooKeeperConnectionTimeout)
+      .concat(Source.single(LeadershipState.Standby(None)))
+      .via(EnrichedFlow.dedup(LeadershipState.Standby(None)))
+      .map { e =>
+        e match {
+          case LeadershipState.ElectedAsLeader =>
+            logger.info("Leader won.")
+          case LeadershipState.Standby(currentLeader) =>
+            logger.info(s"Leader defeated. Current leader: ${currentLeader.getOrElse("-")}")
+        }
+        e
+      }
+  }
+
+  class CuratorEventEmitter(
```

Comment:
I have an idea of representing this event emitter as a `Source[LeadershipState, Cancellable]`. Let's discuss this.

Reply:
It already is a `Source[LeadershipState, Cancellable]`; see the `def apply` method on line 34?

I marked this class private. Ain't nobody got no business using it.